### DISCUSSION TITLE
Add Multi-Vendor feature

### DISCUSSION THREAD

**Discussion by docimin** [medusajs/medusa#2769] [Feature Requests]
Add the ability to allow (Multi-)Vendors to sell Digital & Physical products, the ability for them to create shipment tracking (DHL, USPS etc.), as well as automatic transfers (payments) through Stripe Connect & PayPal.

**Comment by jonebarker** [comment]
Also interested in a multi-vendor solution for digital products.

**Comment by zaheeratwork** [comment]
This is the only show-stopper for us that is stopping or making us nervous to use Medusa. Is there any other open-source alternatives?

**Reply by coopbri** [reply]
Wanted to add to this with my own experience, +1 on Vendure, ~~MIT FOSS~~ (GPLv3, see below 😔) & multi-vendor out of the box :tada: 

https://docs.vendure.io/guides/how-to/multi-vendor-marketplaces/

**Reply by AymenElmehdhaoui** [reply]
vendure license changed from MIT to GPLv3 [(https://github.com/vendure-ecommerce/vendure/pull/2946)](https://github.com/vendure-ecommerce/vendure/pull/2946)

**Reply by ambigos1** [reply]
How is your experience using Vendure so far?
Looking for a multi-vendor solution and I am really want to work with Medusa because of the community but it is not a simple task to implement multi-vendor with admins and roles for each store with Medusa.

@coopbri @AymenElmehdhaoui

**Reply by coopbri** [reply]
> How is your experience using Vendure so far? Looking for a multi-vendor solution and I am really want to work with Medusa because of the community but it is not a simple task to implement multi-vendor with admins and roles for each store with Medusa.
> 
> @coopbri @AymenElmehdhaoui

This is really funny timing. Actually, I ended up finding too many things in Vendure that I really wanted, for example standards such as Relay spec implementations (https://github.com/vendure-ecommerce/vendure/issues/1594) (not an issue for Medusa, since GraphQL isn't currently a first-class citizen, but it is first-class with Vendure). But the thing that most importantly caused me to want to switch off is that they switched from MIT to a copyleft license (GPLv3) as @AymenElmehdhaoui mentioned above (thank you for the correction Aymen). Overall, I ended up beginning to migrate back to Medusa 2 days ago, so it's funny you asked me just now! **The Vendure community is excellent**, copyleft licenses are just **usually** a hard no from me, I prefer fully open source.

At the end of the day, Vendure is an amazing product and I had a good experience with it. The community is very friendly and active. I gave both Medusa and Vendure honest tries (many months), and I am not opposed to switching back (or switching to another ecosystem entirely) if there is a compelling reason to.

By the way, I still use Shopify for some client projects. Figured I'd throw that in the mix in case you ever want to discuss Shopify vs. Medusa, Vendure, ... The hope is to switch off Shopify and others to fully FOSS in the future, but we're not quite there yet.

**EDIT:** Vendure is planning on having a dual license with e.g. MIT for the plugin system ([Vendure Discord discussion](https://discord.com/channels/1100672177260478564/1304794998625271809/1304843180503732376)), but the core is still GPLv3. All of my thoughts above remain, I would not contribute to a copyleft core myself and would rather avoid using it. Let's hope Medusa stays MIT!

**Reply by AymenElmehdhaoui** [reply]
We prefer Vendure for the architecture/technical style: plugins, Typescript, GraphQl, etc. However, we prefer fully open source code more. We choosed the best of both as we took [Vendure v2.x](https://github.com/vendure-ecommerce/vendure/tree/v2.x) (under MIT license, not the latest version). The catch is that we built our solution on top of it and we have to maintain it. Medusajs is still a good choice. I recommend checking [mercurjs](https://github.com/mercurjs/mercur) as it supports marketplace features and is built on Medusajs v1. I checked the repo lately, they are trying to adopt Medusajs v2. As of today, it is under dev. You can find it [here](https://github.com/mercurjs/mercur/tree/v2). Madusa v2 has doc section for multi-vendor marketplace. I haven't checked it in detail yet. You can find it [here](https://docs.medusajs.com/resources/recipes/marketplace)

I believe there is a misconception about GPLv3. It doesn't force you to publish your code open source. If you **prefer to propagate** the code (sell it, copy it, publish it as open source ....), it forces you to propagate it under the same license. You can use/modify Vendure under GPLv3 for personal and commercial reasons without contributing to the project or sharing your code. More details [here](https://vendure.io/blog/2024/06/busting-the-myth-of-gpl)

**Reply by SalahAdDin** [reply]
> We prefer Vendure for the architecture/technical style: plugins, Typescript, GraphQl, etc. However, we prefer fully open source code more. We choosed the best of both as we took [Vendure v2.x](https://github.com/vendure-ecommerce/vendure/tree/v2.x) (under MIT license, not the latest version). The catch is that we built our solution on top of it and we have to maintain it. Medusajs is still a good choice. I recommend checking [mercurjs](https://github.com/mercurjs/mercur) as it supports marketplace features and is built on Medusajs v1. I checked the repo lately, they are trying to adopt Medusajs v2. As of today, it is under dev. You can find it [here](https://github.com/mercurjs/mercur/tree/v2). Madusa v2 has doc section for multi-vendor marketplace. I haven't checked it in detail yet. You can find it [here](https://docs.medusajs.com/resources/recipes/marketplace)
> 
> I believe there is a misconception about GPLv3. It doesn't force you to publish your code open source. If you **prefer to propagate** the code (sell it, copy it, publish it as open source ....), it forces you to propagate it under the same license. You can use/modify Vendure under GPLv3 for personal and commercial reasons without contributing to the project or sharing your code. More details [here](https://vendure.io/blog/2024/06/busting-the-myth-of-gpl)

Yet Medusa has no a good RBAC plugin accessible for all developers.

**Reply by coopbri** [reply]
I ended up forking Vendure v2 (MIT). Rest assured, this fork will forever be MIT, even if it becomes archived. **There will be no transition to a copyleft license.**

If anyone is interested, I called this fork "Halo", https://github.com/coopbri/halo

I am already trialing in production, but please do not yet use it in production yourself as there are no published packages, still traces of the "Vendure" name, etc. If interested in this project, please use Medusa, Vendure, Shopify, or another ecommerce platform, at least for now, but keep an eye on it. I want to make sure it is stable before anyone uses it in any production capacity

I made this fork solely based on my disagreement with the MIT to GPLv3 migration. With that said, I love Vendure upstream and the community, and deeply appreciate the work that has been put into it.

**Important:** I am talking with some community members who also had Vendure forks. If theirs is farther along, I will contribute to those instead and clarify that here ASAP. I am just trying to do what's most open for the community

Again, **do not use this in production**, this is currently just an experiment. So far I have replaced the main build system with Bun and removed some plugins/packages, not much else. I am currently experimenting with a Rust-powered job queue and a WASM-powered plugin system.

**Reply by coopbri** [reply]
> > We prefer Vendure for the architecture/technical style: plugins, Typescript, GraphQl, etc. However, we prefer fully open source code more. We choosed the best of both as we took [Vendure v2.x](https://github.com/vendure-ecommerce/vendure/tree/v2.x) (under MIT license, not the latest version). The catch is that we built our solution on top of it and we have to maintain it. Medusajs is still a good choice. I recommend checking [mercurjs](https://github.com/mercurjs/mercur) as it supports marketplace features and is built on Medusajs v1. I checked the repo lately, they are trying to adopt Medusajs v2. As of today, it is under dev. You can find it [here](https://github.com/mercurjs/mercur/tree/v2). Madusa v2 has doc section for multi-vendor marketplace. I haven't checked it in detail yet. You can find it [here](https://docs.medusajs.com/resources/recipes/marketplace)
> > I believe there is a misconception about GPLv3. It doesn't force you to publish your code open source. If you **prefer to propagate** the code (sell it, copy it, publish it as open source ....), it forces you to propagate it under the same license. You can use/modify Vendure under GPLv3 for personal and commercial reasons without contributing to the project or sharing your code. More details [here](https://vendure.io/blog/2024/06/busting-the-myth-of-gpl)
> 
> Yet Medusa has no a good RBAC plugin accessible for all developers.

Found this active discussion about RBAC: https://github.com/medusajs/medusa/discussions/9909

**Reply by SalahAdDin** [reply]
> > > We prefer Vendure for the architecture/technical style: plugins, Typescript, GraphQl, etc. However, we prefer fully open source code more. We choosed the best of both as we took [Vendure v2.x](https://github.com/vendure-ecommerce/vendure/tree/v2.x) (under MIT license, not the latest version). The catch is that we built our solution on top of it and we have to maintain it. Medusajs is still a good choice. I recommend checking [mercurjs](https://github.com/mercurjs/mercur) as it supports marketplace features and is built on Medusajs v1. I checked the repo lately, they are trying to adopt Medusajs v2. As of today, it is under dev. You can find it [here](https://github.com/mercurjs/mercur/tree/v2). Madusa v2 has doc section for multi-vendor marketplace. I haven't checked it in detail yet. You can find it [here](https://docs.medusajs.com/resources/recipes/marketplace)
> > > I believe there is a misconception about GPLv3. It doesn't force you to publish your code open source. If you **prefer to propagate** the code (sell it, copy it, publish it as open source ....), it forces you to propagate it under the same license. You can use/modify Vendure under GPLv3 for personal and commercial reasons without contributing to the project or sharing your code. More details [here](https://vendure.io/blog/2024/06/busting-the-myth-of-gpl)
> > 
> > 
> > Yet Medusa has no a good RBAC plugin accessible for all developers.
> 
> Found this active discussion about RBAC: #9909

No any answer from the Medusa's team.

**Reply by coopbri** [reply]
Yes, I am just sharing an active discussion that others can track about that issue. I did see you already have been communicating in that issue though.

**Comment by isneverdead** [comment]
is there anyone create multi vendor store using the newest version of medusa? 
i saw there is some article about this but i can't follow it because they using old medusa version

[UPDATE]
hi guys, i found this great article may can help you to create multi vendor
[https://medusajs.com/blog/extending-medusa-usecase-marketplace/](https://medusajs.com/blog/extending-medusa-usecase-marketplace/)

**Reply by Feng-Jiang28** [reply]
I am also trying to create a multi-vendor project that works like ebay. Can we discuss that?

**Comment by thejspro** [comment]
This is a critical blocker for us using Medusa. We need something working out of the box, without the need to customise the backend. It's the single, most lacking feature of Medusa.

**Comment by vitalijalbu** [comment]
Also want to create a marketplace, where all users can sell or buy products

**Comment by oliwerhelsen** [comment]
I think this blogpost is a good start. Helped me a lot and resolve our needs for a marketplace.
https://medusajs.com/blog/extending-medusa-usecase-marketplace/

**Reply by kleenkanteen** [reply]
@oliwerhelsen How would you give users access to the admin dashboard just for their products?

**Reply by herawais** [reply]
@kleenkanteen did you find the answer ?

**Reply by kleenkanteen** [reply]
@herawais I did not

**Reply by oliwerhelsen** [reply]
Sorry for late reply guys. We're using subdomains as a look up for tenant by the tenantId and by that we can then control everything within the dashboard to just handle the tenant related stuff.

**Reply by th-devangnaghera** [reply]
@oliwerhelsen How do you guys manage payments for each tenant?

**Reply by oliwerhelsen** [reply]
We made our own solution in the storefront and then take use of the events.

**Reply by SalahAdDin** [reply]
@oliwerhelsen Thinking on a plugin for it?

**Reply by ambigos1** [reply]
@oliwerhelsen It Sounds like a smart solution to use a subdomain.
Are you using a single DB instance for all of your tenants?
as a super-admin how can you see all sales and products of all of your tenants?

**Comment by SGFGOV** [comment]
You need to overrirde the retrieve function

On Tue, 13 Jun, 2023, 11:42 Muhammad Awais, ***@***.***>
wrote:

> @kleenkanteen <https://github.com/kleenkanteen> did you find the answer ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/medusajs/medusa/discussions/2769#discussioncomment-6159618>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXEQJHHN2WS5E26VIJD4LY3XLAAETANCNFSM6AAAAAAS35CYII>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

-- 

This message (including any attachments) may contain confidential, 
proprietary, privileged and/or private
information. The information is 
intended to be for the use of the individual or entity designated above. If
you are not the intended recipient of this message, please notify the 
sender immediately, and delete the
message and any attachments. Any 
disclosure, reproduction, distribution or other use of this message or
any 
attachments by an individual or entity other than the intended recipient is 
prohibited.

**Reply by kleenkanteen** [reply]
@SGFGOV Are you replying to me? Can you elaborate?

**Reply by SGFGOV** [reply]
Well you need to modify the productservice retrieve/list functions to give a different set of products when a authenticated user logs in

You can check if the loggedInUser exists and conditionally return the products you are interested in

**Reply by kleenkanteen** [reply]
@SGFGOV When you say checking if loggedInUser exists, is that that the same as what is done in the "Create Middleware to Register Logged-In User" section in this [guide](https://medusajs.com/blog/extending-medusa-usecase-marketplace/)?

**Reply by kleenkanteen** [reply]
@SGFGOV Any update on above comment?

**Reply by SGFGOV** [reply]
Yes

**Comment by vitalijalbu** [comment]
HI guys, I'd prefer going with some headless cms instead of medusa, then will develop stripe checkout manually.

**Comment by alexandrosk** [comment]
I think anyone can build this with a plugin now let's see :)

**Comment by Kiwifu** [comment]
Also interested in a multi-vendor solution for multi-categories

**Comment by agnelnieves** [comment]
https://docs.medusajs.com/recipes/marketplace#associate-entities-with-stores

**Reply by SalahAdDin** [reply]
So we can create a plugin for multi-vendor following this documentation, right?

**Comment by dealwith** [comment]
It would be really nice, to have the multi vendor feature out of the box, without configure and adapting backend models and backend services. Current guides that does exists on medusa are really helpful but the job to make the real multi-store commerce is harder. I would really appreciate multi-store feature to be in a roadmap

**Comment by adrian09h** [comment]
I am also interested in the multi-vendor solution.
I am willing to pay to buy multi-veondor extension or plugin.

**Comment by SGFGOV** [comment]
I've built a plugin for creating a marketplace and willing to opensource it, 
Things it currently doesn't have
1. Tests. Any CI and CD pipe line will need this, as future fixes and enhancements shouldn't accidentally break current features 
2.  Maintenance support.  Right now it's built to meet my requirements. However, to make it more generic will need some development efforts
4.  Documentation. Since I wrote it, i know how it works, but if other want to use it , they will need documentation and how to guides.
5. A full-fledged vendor management interface. Right now its sort of make shift using the medusa admin, though the backend is setup for multivendor, the front end needs some work, particularly with respect to things like permissions. 

Doing all this will take some time, effort and needless to add, funding. If the community is willing to help raise the funding via crowdsourcing, I'm willing to open source and maintain this plugin for as long as the funding exists. I am sure it will help thousands of developers. If there are folks who feel this is viable, I can put together a work and funding estimate. 

Also if there are any other features you require, put it as a part of the comments. I'll include it in the estimate

**Reply by SalahAdDin** [reply]
Multi vendor?

**Reply by SGFGOV** [reply]
Yes

**Reply by SalahAdDin** [reply]
what do you want to use for documentation?

**Reply by SGFGOV** [reply]
let me know what you think will be best for documentation

**Reply by SalahAdDin** [reply]
It was learning to work with docusaurus, but I saw many companies writing their documentation with NextJS.

**Reply by haf** [reply]
@SGFGOV Did you open source it?

**Reply by akordzik** [reply]
Hi @SGFGOV, are you still willing to open source it?

**Reply by SGFGOV** [reply]
It's all v1 so.  Am not sure if it's relavent now

**Comment by kleenkanteen** [comment]


**Reply by SalahAdDin** [reply]
for multivendor?

**Reply by kleenkanteen** [reply]
@SalahAdDin yes. sorry didnt reply in the thread as did it thruogh email. 

wdym how btw? it uses markdown which is easy to use for documetnatoin

**Reply by cbou** [reply]
Hono & Vitepress aren't Multi Vendor Marketplace, not even ecommerce solution...

**Comment by kleenkanteen** [comment]


**Reply by SalahAdDin** [reply]
How?

**Comment by SGFGOV** [comment]
Looks like there is some interest, Trying to see if I can get at least 25 up votes to this suggestion

**Reply by cbou** [reply]
Would be really nice!

**Comment by adamlesniak** [comment]
Interesting implementation using `nx` might be good if Medusa grows long term: (Might feel a bit chore like with all the CLI and scripts)

https://github.com/Keith-Hon/medusa-marketplace/tree/master/apps

I'm looking into some solution myself by extending current repo https://github.com/adamlesniak/medusa-marketplace - took inspiration by @shahednasser (https://github.com/shahednasser/medusa-1.8-marketplace-tutorial) - thinking it could be nice to wrap it onto dedicated module and make it plug-n-play like other Medusa modules.

**Reply by SalahAdDin** [reply]
What is this experiment for? can you explain it, please?

**Reply by adamlesniak** [reply]
As I understand based on the original post - marketplace will provide multi vendor ability - where vendor/business can register, create their own store and facility to sell products - along with customisation such as shipping, pricing tax etc...

**Reply by SalahAdDin** [reply]
Where is the original post? Amazing!

**Reply by adamlesniak** [reply]
Original post is first post in this discussion is what I mean

**Reply by SalahAdDin** [reply]
I refers to the new recipe: https://docs.medusajs.com/recipes/marketplace
hehehehe

**Comment by adevinwild** [comment]
**Hello everyone!**
For those of you just passing through, at the moment of this message, there is some solutions to help you build a marketplace / multivendor project with Medusa :

## perseides - Learn how to build your marketplace with Medusa (`V1.x` only)
I've started a series of articles on how to build a marketplace from scratch using Medusa and Stripe Connect.
They cover a range of concepts and teach you how to customize Medusa from the inside to build a functional marketplace.

You can find more information here: [perseides.org](https://perseides.org/)

## mercurjs - Medusa MultiVendor Marketplace Accelerator (`V1.x` and `V2`)
Mercur, is a plug-n-play package built by @gtomaka that lets you directly install a backend with multi-vendor functionalities, as well as a storefront and an admin UI!

You can find more information here: [mercurjs.com](https://mercurjs.com/)

**Reply by akordzik** [reply]
Sweet, unfortunately it's not compatible with V2

https://github.com/mercurjs/mercur/issues/31

**Reply by adevinwild** [reply]
Hey @akordzik 
Just wanted to make a quick update here, @gtomaka have commented the issue here : https://github.com/mercurjs/mercur/issues/31

Mercur will be compatible for V2 🙌🏼

**Reply by coopbri** [reply]
This is sick! Thanks @adevinwild, you're the man

**Reply by ambigos1** [reply]
@coopbri, it seems like you have vast experience in e-commerce libraries.
I hope to hear your opinion on Mercurjs in the future :)

**Reply by sianami123** [reply]
can it one click deploy on coolify?

**Comment by darkzbaron** [comment]
+1 here
